src/dbnode/storage/throttler.go

```diff
@@ -0,0 +1,126 @@
+package storage
+
+import (
+	"sync"
+)
+
+// Throttler controls fair access to limited resources.
+type Throttler struct {
+	sync.Mutex
+
+	keyState map[string]*keyContext
+	// Each entry in the queue should be unique to avoid unfair access
+	keyQueue []string
+
+	globalCurrentWeight int
+	globalMaxWeight     int
+}
+
+type keyContext struct {
+	// Requests for weight to be granted which are waiting.
+	waiting []request
+	// Currently granted weight for this key.
+	currentWeight int
+}
+
+type request struct {
+	blockCh chan struct{}
+	weight  int
+}
+
+// Acquire blocks until the requested weight is granted to the specified key.
+func (t *Throttler) Acquire(key string, weight int) {
+	t.Lock()
+	defer t.Unlock()
+
+	maxWeightPerKey := t.maxWeightPerKey()
+
+	currentKey, alreadyExists := t.keyState[key]
+	if !alreadyExists {
+		t.keyState[key] = &keyContext{
+			currentWeight: 0,
+			waiting:       make([]request, 0),
+		}
+	}
+
+	// If new weight would keep the key under the per-key limit then grant.
+	// Otherwise enqueue and block.
+	if currentKey.currentWeight+weight < maxWeightPerKey {
+		currentKey.currentWeight += weight
+		t.globalCurrentWeight += weight
+	} else {
+		blockCh := make(chan struct{}, 0)
+		currentKey.waiting = append(currentKey.waiting, request{
+			blockCh: blockCh,
+			weight:  weight,
+		})
+
+		// If this is first request by the key, then enqueue it for releasing.
+		if !alreadyExists {
+			t.keyQueue = append(t.keyQueue, key)
+		}
+
+		// Block until enough weight is released.
+		<-blockCh
+	}
+}
+
+// Release frees the weight associated with the key to be available for others.
+func (t *Throttler) Release(key string, weight int) {
+	t.Lock()
+	defer t.Unlock()
+
+	currentKey := t.keyState[key]
+
+	// Reduce granted weight associated with this key.
+	currentKey.currentWeight -= weight
+	t.globalCurrentWeight -= weight
+
+	// calculate dynamic limit
+	maxWeightPerKey := t.maxWeightPerKey()
+
+	// Cycle through the queue of keys waiting for resources to determine
```

Comment:
unless I'm reading this wrong, lots of "light" activity will starve "heavy" waiters - if truly striving for fairness might need a policy that defers granting access to a "light" waiter if there is a "heavy" waiter that has been blocked for some period of time.

Reply:
Hm yeah you're right but that's not intentional. We want to prevent starvation in any case, and the idea is more to just make sure all N concurrent users get the same share of bandwidth as we churn through everyone who is waiting. The logic becomes simpler w/o the whole `weight` concept, and after we looked more into the intended call sites it seems unnecessary, so I think I'll just remove it which should solve this problem.